File: examples/flutter/cases/tasks/task_move/task_move.ui.case.dart
import 'package:flutter/material.dart';

import '../../../core/ui.case.dart';
import 'task_move.domain.case.dart';

typedef TaskMovedCallback = void Function(Map<String, dynamic> task);

class TaskMoveUi extends BaseUiCase<UIState> {
  TaskMoveUi(UiContext ctx)
    : super(ctx, <String, dynamic>{'loading': false, 'error': null});

  final TaskMoveDomain _domainCase = TaskMoveDomain();
  bool _moveLocked = false;

  @override
  Widget view() {
    final designSystem = _resolveDesignSystem();
    final task = ctx.extra?['task'];

    if (task is! Map<String, dynamic>) {
      throw Exception('task_move.ui requires extra.task');
    }

    return _TaskMoveView(uiCase: this, designSystem: designSystem, task: task);
  }

  @override
  Future<void> test() async {
    view();
    final task = ctx.extra?['task'];
    if (task is! Map<String, dynamic>) {
      throw Exception('test: task_move.ui requires extra.task');
    }

    final result = await _service(
      TaskMoveInput(taskId: task['id'], targetStatus: 'doing'),
    );

    if (result.task.status != 'doing') {
      throw Exception('test: ui service must return the moved task');
    }

    var threw = false;
    try {
      await _service(const TaskMoveInput(taskId: '', targetStatus: 'done'));
    } catch (_) {
      threw = true;
    }

    if (!threw) {
      throw Exception('test: ui service must reject invalid input');
    }

    if (!_acquireMoveLock()) {
      throw Exception('test: first move lock acquisition must succeed');
    }

    if (_acquireMoveLock()) {
      throw Exception('test: move lock must reject reentry');
    }

    _releaseMoveLock();

    if (!_acquireMoveLock()) {
      throw Exception('test: move lock must be releasable');
    }

    _releaseMoveLock();
  }

  dynamic _resolveDesignSystem() {
    final packages = ctx.packages;
    final designSystem = packages?['designSystem'];

    if (designSystem == null) {
      throw Exception('task_move.ui requires packages.designSystem');
    }

    return designSystem;
  }

  Future<TaskMoveOutput> _service(TaskMoveInput input) async {
    _domainCase.validate(input);
    return _repository(input);
  }

  Future<TaskMoveOutput> _repository(TaskMoveInput input) async {
    final api = ctx.api;
    if (api == null) {
      throw Exception('task_move.ui requires ctx.api');
    }

    final response = await api.request(<String, dynamic>{
      'method': 'PATCH',
      'url': '/tasks/${input.taskId}/status',
      'body': <String, dynamic>{'targetStatus': input.targetStatus},
    });

    final result = TaskMoveOutput.fromJson(response);
    if (result.task.id.isEmpty) {
      throw Exception('task_move.ui received an invalid move response');
    }

    return result;
  }

  bool _acquireMoveLock() {
    if (_moveLocked) {
      return false;
    }

    _moveLocked = true;
    return true;
  }

  void _releaseMoveLock() {
    _moveLocked = false;
  }
}

class _TaskMoveView extends StatefulWidget {
  const _TaskMoveView({
    required this.uiCase,
    required this.designSystem,
    required this.task,
  });

  final TaskMoveUi uiCase;
  final dynamic designSystem;
  final Map<String, dynamic> task;

  @override
  State<_TaskMoveView> createState() => _TaskMoveViewState();
}

class _TaskMoveViewState extends State<_TaskMoveView> {
  late Map<String, dynamic> _state;

  @override
  void initState() {
    super.initState();
    _state = Map<String, dynamic>.from(widget.uiCase.state);
  }

  Future<void> _move(String nextStatus) async {
    if (!widget.uiCase._acquireMoveLock()) {
      return;
    }

    setState(() {
      _state = <String, dynamic>{'loading': true, 'error': null};
    });

    try {
      final result = await widget.uiCase._service(
        TaskMoveInput(taskId: widget.task['id'], targetStatus: nextStatus),
      );

      final callback = widget.uiCase.ctx.extra?['onTaskMoved'];
      if (callback is TaskMovedCallback) {
        callback(result.task.toJson());
      }

      if (!mounted) {
        return;
      }

      setState(() {
        _state = <String, dynamic>{'loading': false, 'error': null};
      });
    } catch (error) {
      if (!mounted) {
        return;
      }

      setState(() {
        _state = <String, dynamic>{
          'loading': false,
          'error': error.toString().replaceFirst('Exception: ', ''),
        };
      });
    } finally {
      widget.uiCase._releaseMoveLock();
    }
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: <Widget>[
        widget.designSystem.moveTaskAction(
          currentStatus: widget.task['status']?.toString() ?? 'todo',
          submitting: _state['loading'] == true,
          onMove: _move,
        ),
        if (_state['error'] != null) ...<Widget>[
          const SizedBox(height: 8),
          Text(
            _state['error']?.toString() ?? '',
            style: const TextStyle(color: Color(0xFF9A2D1F), fontSize: 13),
          ),
        ],
      ],
    );
  }
}

File: examples/angular/cases/tasks/task_move/task_move.ui.case.ts
import { CommonModule, NgComponentOutlet } from "@angular/common";
import { Component, Input, Type, signal } from "@angular/core";

import { BaseUiCase, type UIState, type UiContext } from "../../../core/ui.case";
import { type AppHttpClient } from "../../../core/shared/app_infra_contracts";
import {
  TaskMoveDomain,
  type Task,
  type TaskMoveInput,
  type TaskMoveOutput,
} from "./task_move.domain.case";

type ViewState = UIState & {
  loading: boolean;
  error: string | null;
};

interface AngularViewUnit {
  component: Type<unknown>;
  inputs?: Record<string, unknown>;
}

interface TaskMoveDesignSystem {
  MoveTaskAction: Type<unknown>;
}

interface UiExtraShape {
  task?: Task;
  onTaskMoved?: (task: Task) => void;
}

@Component({
  selector: "app-task-move-view",
  standalone: true,
  imports: [CommonModule, NgComponentOutlet],
  template: `
    <div>
      <ng-container
        [ngComponentOutlet]="designSystem.MoveTaskAction"
        [ngComponentOutletInputs]="moveActionInputs"
      />

      <div
        *ngIf="state().error as error"
        style="color: #9a2d1f; font-size: 0.82rem; margin-top: 0.55rem;"
      >
        {{ error }}
      </div>
    </div>
  `,
})
class TaskMoveViewComponent {
  @Input({ required: true }) caseRef!: TaskMoveUi;
  @Input({ required: true }) designSystem!: TaskMoveDesignSystem;
  @Input({ required: true }) task!: Task;

  protected readonly state = signal<ViewState>({
    loading: false,
    error: null,
  });

  protected get moveActionInputs(): Record<string, unknown> {
    const state = this.state();

    return {
      currentStatus: this.task.status,
      submitting: state.loading,
      onMove: (nextStatus: Task["status"]) => {
        void this.move(nextStatus);
      },
    };
  }

  private async move(nextStatus: Task["status"]): Promise<void> {
    if (!this.caseRef.acquireMoveLock()) {
      return;
    }

    this.state.set({
      loading: true,
      error: null,
    });

    try {
      const result = await this.caseRef.runService({
        taskId: this.task.id,
        targetStatus: nextStatus,
      });

      this.caseRef.onTaskMoved(result.task);

      this.state.set({
        loading: false,
        error: null,
      });
    } catch (error: unknown) {
      this.state.set({
        loading: false,
        error:
          error instanceof Error ? error.message : "Failed to move task",
      });
    } finally {
      this.caseRef.releaseMoveLock();
    }
  }
}

export class TaskMoveUi extends BaseUiCase<ViewState> {
  private readonly domainCase = new TaskMoveDomain();
  private moveLocked = false;

  constructor(ctx: UiContext) {
    super(ctx, {
      loading: false,
      error: null,
    });
  }

  public view(): AngularViewUnit {
    const packages = this.ctx.packages as { designSystem?: TaskMoveDesignSystem } | undefined;
    const DesignSystem = packages?.designSystem;
    const extra = this.ctx.extra as UiExtraShape | undefined;

    if (!DesignSystem) {
      throw new Error("task_move.ui requires packages.designSystem");
    }

    if (!extra?.task) {
      throw new Error("task_move.ui requires extra.task");
    }

    return {
      component: TaskMoveViewComponent,
      inputs: {
        caseRef: this,
        designSystem: DesignSystem,
        task: extra.task,
      },
    };
  }

  public async test(): Promise<void> {
    const view = this.view();
    if (!view) {
      throw new Error("test: view must return a visual unit");
    }

    const result = await this._service({
      taskId: "task_001",
      targetStatus: "doing",
    });

    if (result.task.status !== "doing") {
      throw new Error("test: ui service must return the moved task");
    }

    let threw = false;
    try {
      await this._service({
        taskId: "",
        targetStatus: "done",
      });
    } catch {
      threw = true;
    }

    if (!threw) {
      throw new Error("test: ui service must reject invalid input");
    }

    if (!this._acquireMoveLock()) {
      throw new Error("test: first move lock acquisition must succeed");
    }

    if (this._acquireMoveLock()) {
      throw new Error("test: move lock must reject reentry");
    }

    this._releaseMoveLock();

    if (!this._acquireMoveLock()) {
      throw new Error("test: move lock must be releasable");
    }

    this._releaseMoveLock();
  }

  public async runService(input: TaskMoveInput): Promise<TaskMoveOutput> {
    return this._service(input);
  }

  public onTaskMoved(task: Task): void {
    const extra = this.ctx.extra as UiExtraShape | undefined;
    extra?.onTaskMoved?.(task);
  }

  public acquireMoveLock(): boolean {
    return this._acquireMoveLock();
  }

  public releaseMoveLock(): void {
    this._releaseMoveLock();
  }

  protected async _service(input: TaskMoveInput): Promise<TaskMoveOutput> {
    this.domainCase.validate(input);
    return this._repository(input);
  }

  protected async _repository(
    input: TaskMoveInput
  ): Promise<TaskMoveOutput> {
    const response = await this.resolveApiClient().request({
      method: "PATCH",
      url: `/tasks/${input.taskId}/status`,
      body: {
        targetStatus: input.targetStatus,
      },
    });

    const result = response as TaskMoveOutput;
    this.domainCase.validateOutput(result);
    return result;
  }

  protected _acquireMoveLock(): boolean {
    if (this.moveLocked) {
      return false;
    }

    this.moveLocked = true;
    return true;
  }

  protected _releaseMoveLock(): void {
    this.moveLocked = false;
  }

  private resolveApiClient(): AppHttpClient {
    if (!this.ctx.api) {
      throw new Error("task_move.ui requires ctx.api");
    }

    return this.ctx.api;
  }
}
